### PR TITLE
Android: Implement retention experiment to treat returning users differently

### PR DESCRIPTION
**Pull request by anikiki**
Task/Issue URL: https://app.asana.com/0/0/1201408441367239/f

### Description
This adds a 3-way experiment (control and 2 variants). Videos for all versions and exact details can be found in https://app.asana.com/0/1142021229838617/1201393848577545/f

### Steps to test this PR

**_Control_**
_**Code changes before testing**: Change the weighting of all variants in the VariantManager to 0.0 and keep the weight of "zk" to 1.0_

1. Freshly install the app.
2. On the Welcome page you should see the normal welcome screen:
- text: _The Internet can be kinda creepy. Not to worry! Searching and browsing privately is easier than you think._
- only one button with the label: _Let’s Do It!_
3. Tap on Let’s Do It! button (Pixel: _m_odp_p_)
4. You should see the normal onboarding.

control

**_Variant 1 - New users path_**
_**Code changes before testing**: Change the weighting of all variants in the VariantManager to 0.0 and keep the weight of "zv" to 1.0_

1. Freshly install the app.
2. On the Welcome page you should see a welcome screen with a text and two buttons:
- text: _The Internet can be kinda creepy. Not to worry! Searching and browsing privately is easier than you think. Let’s get started 👍_
- two buttons with the labels: _I’m new!_ and _I’ve been here before._
3. Tap on I’m new! button (Pixel: _m_odn_p_)
4. You should see the normal onboarding.

variant1

**_Variant 1 - Returning users path_**
_**Code changes before testing**: Change the weighting of all variants in the VariantManager to 0.0 and keep the weight of "zv" to 1.0_

1. Freshly install the app.
2. On the Welcome page you should see a welcome screen with a text and two buttons:
- text: _The Internet can be kinda creepy. Not to worry! Searching and browsing privately is easier than you think. Let’s get started 👍_
- two buttons with the labels: _I’m new!_ and _I’ve been here before._
3. Tap on I’ve been here before. button (Pixel: _m_odr_p_)
4. You should see the default browser check and no other dialogs afterwards.

**_Variant 2 - New users path_**
_**Code changes before testing**: Change the weighting of all variants in the VariantManager to 0.0 and keep the weight of "zz" to 1.0_

1. Freshly install the app.
2. On the Welcome page you should see a welcome screen with a text and two buttons:
- text: _The Internet can be kinda creepy. Not to worry! Searching and browsing privately is easier than you think. Let’s get started 👍_
- two buttons with the labels: _I’m new!_ and _I’ve been here before._
3. Tap on I’m new! button (Pixel: _m_odn_p_)
4. You should see the normal onboarding.

variant2

**_Variant 2 - Returning users path_**
_**Code changes before testing**: Change the weighting of all variants in the VariantManager to 0.0 and keep the weight of "zz" to 1.0_

1. Freshly install the app.
2. On the Welcome page you should see a welcome screen with a text and two buttons:
- text: _The Internet can be kinda creepy. Not to worry! Searching and browsing privately is easier than you think. Let’s get started 👍_
- two buttons with the labels: _I’m new!_ and _I’ve been here before._
3. Tap on I’ve been here before. button (Pixel: _m_odr_p_)
4. You should see the default browser check and then a prompt to add the search widget. 
- widget prompt title: Add the DuckDuckGo widget to your home screen
- widget prompt buttons: No Thanks and Add Widget 
- notice no Dax logo on the background while the widget prompt is shown.
- no other dialogs shown after this step
- the Dax logo is shown on empty new tab.

widget

### CODE REVIEW

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt [hunk 1]
```diff
@@ -248,6 +250,9 @@ class BrowserTabFragment :
     @Inject
     lateinit var themingDataStore: ThemingDataStore
 
+    @Inject
+    lateinit var onboardingStore: OnboardingStore
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt:254]: I think this is not used here.

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt [hunk 2]
```diff
@@ -2259,7 +2266,7 @@ class BrowserTabFragment :
 
         private fun showHomeBackground(favorites: List<FavoritesQuickAccessAdapter.QuickAccessFavorite>) {
             if (favorites.isEmpty()) {
-                homeBackgroundLogo.showLogo()
+                if (hideDaxBackgroundLogo) homeBackgroundLogo.hideLogo() else homeBackgroundLogo.showLogo()
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt:2269]: Could `hideDaxBackgroundLogo` be replaced by a method param? If I'm not wrong, we can decide if we need to show the logo or not based on the emitted `CtaViewState`.

(I'm just trying to remove any global state if possible)

File: app/src/main/java/com/duckduckgo/app/cta/ui/Cta.kt [hunk 3]
```diff
@@ -371,9 +371,9 @@ sealed class HomePanelCta(
 ) : Cta, ViewCta {
 
     override fun showCta(view: View) {
-        view.ctaIcon.setImageResource(image)
-        view.ctaTitle.text = view.context.getString(title)
-        view.ctaSubtitle.text = view.context.getString(description)
+        view.ctaIcon?.setImageResource(image)
```
Comment (cmonfortep) [inline app/src/main/java/com/duckduckgo/app/cta/ui/Cta.kt:374]: Q: did you see any issue here that made you use `?`?

File: app/src/androidTest/java/com/duckduckgo/app/cta/ui/CtaViewModelTest.kt [hunk 4]
```diff
@@ -695,6 +695,32 @@ class CtaViewModelTest {
         assertNull(fireDialogCta)
     }
 
+    @Test
+    fun whenRefreshCtaOnHomeTabAndReturningUsersNoOnboardingEnabledTrueAndWidgetCompatibleThenReturnNull() = coroutineRule.runBlocking {
```
Comment (cmonfortep) [inline app/src/androidTest/java/com/duckduckgo/app/cta/ui/CtaViewModelTest.kt:699]: `AndWidgetCompatible` does this apply to this test?

I think for this test case, we want to highlight that user has not reached the threshold, wdyt?

**Review by cmonfortep — APPROVED** [review]
Nice work @anikiki Tested and it works as described.

Just left some minor comments.

File: app/src/androidTest/java/com/duckduckgo/app/cta/ui/CtaViewModelTest.kt (hunk 4 above)
Comment (anikiki) [reply app/src/androidTest/java/com/duckduckgo/app/cta/ui/CtaViewModelTest.kt:699]: The part "ReturningUsersNoOnboardingEnabledTrue" means we're in the variant where the widget will never be shown, regardless of the threshold.

File: app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt (hunk 1 above)
Comment (anikiki) [reply app/src/main/java/com/duckduckgo/app/browser/BrowserTabFragment.kt]: Good catch, removed.

File: app/src/main/java/com/duckduckgo/app/onboarding/ui/page/WelcomePageViewModel.kt [hunk 5]
```diff
@@ -20,24 +20,51 @@ import android.annotation.SuppressLint
...
     private val appInstallStore: AppInstallStore,
     private val context: Context,
     private val pixel: Pixel,
-    private val defaultRoleBrowserDialog: DefaultRoleBrowserDialog
+    private val defaultRoleBrowserDialog: DefaultRoleBrowserDialog,
+    private val onboardingStore: OnboardingStore,
+    private val variantManager: VariantManager
 ) : ViewModel() {
 
+    sealed class ViewState {
```
[the remark this answers is not in the document]
Comment (anikiki) [reply app/src/main/java/com/duckduckgo/app/onboarding/ui/page/WelcomePageViewModel.kt:44]: I didn't want to unify them and undo the work after. It's easier handle the experiment.

File: app/src/main/java/com/duckduckgo/app/pixels/AppPixelName.kt [hunk 6]
```diff
@@ -39,6 +39,10 @@ enum class AppPixelName(override val pixelName: String) : Pixel.PixelName {
     ONBOARDING_DAX_ALL_CTA_HIDDEN("m_odc_h"),
     ONBOARDING_DAX_CTA_OK_BUTTON("m_odc_ok"),
 
+    ONBOARDING_DAX_PRIMARY_CTA_PRESSED("m_odp_p"),
```
[the remark this answers is not in the document]
Comment (anikiki) [reply app/src/main/java/com/duckduckgo/app/pixels/AppPixelName.kt]: Yes, updated.

File: app/src/main/java/com/duckduckgo/app/onboarding/ui/page/WelcomePageViewModel.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (cmonfortep) [reply app/src/main/java/com/duckduckgo/app/onboarding/ui/page/WelcomePageViewModel.kt:44]: Sound reasonable 👍

File: app/src/androidTest/java/com/duckduckgo/app/cta/ui/CtaViewModelTest.kt (hunk 4 above)
Comment (cmonfortep) [reply app/src/androidTest/java/com/duckduckgo/app/cta/ui/CtaViewModelTest.kt:699]: gotcha, so `AndWidgetCompatible` means that in a normal scenario the widget would appear...

not sure if it's clear enough, but I don't have a better name...

File: app/src/androidTest/java/com/duckduckgo/app/onboarding/ui/page/WelcomePageViewModelTest.kt [hunk 7]
```diff
@@ -167,4 +189,94 @@ class WelcomePageViewModelTest {
...
+        emitShowDialogAndFireAndHideTips()
+    }
+
+    private fun emitShowDialogAndFireAndNeverHideTips() = coroutineRule.runBlocking {
+        whenever(defaultRoleBrowserDialog.shouldShowDialog()).thenReturn(true)
+        val intent = Intent()
+        whenever(defaultRoleBrowserDialog.createIntent(any())).thenReturn(intent)
+
+        val launch = launch {
+            viewEvents.collect { state ->
```
[the remark this answers is not in the document]
Comment (anikiki) [reply app/src/androidTest/java/com/duckduckgo/app/onboarding/ui/page/WelcomePageViewModelTest.kt]: Updated all tests in this file to use turbine.

File: app/src/androidTest/java/com/duckduckgo/app/onboarding/ui/page/WelcomePageViewModelTest.kt [hunk 8]
```diff
@@ -167,4 +189,94 @@ class WelcomePageViewModelTest {
...
+    @Test
+    fun whenOnReturningUserClickedAndReturningUsersNoOnboardingEnabledThenEmitShowDialog() = coroutineRule.runBlocking {
+        whenever(mockVariantManager.getVariant()).thenReturn(VariantManager.ACTIVE_VARIANTS.first { it.key == "zv" })
+        emitShowDialogAndFireAndHideTips()
+    }
+
+    @Test
+    fun whenOnReturningUserClickedAndReturningUsersWidgetPromotionEnabledThenEmitShowDialog() = coroutineRule.runBlocking {
+        whenever(mockVariantManager.getVariant()).thenReturn(VariantManager.ACTIVE_VARIANTS.first { it.key == "zz" })
+        emitShowDialogAndFireAndHideTips()
```
[the remark this answers is not in the document]
Comment (anikiki) [reply app/src/androidTest/java/com/duckduckgo/app/onboarding/ui/page/WelcomePageViewModelTest.kt]: Updated the tests.

File: app/src/androidTest/java/com/duckduckgo/app/onboarding/ui/page/WelcomePageViewModelTest.kt [hunk 9]
```diff
@@ -167,4 +189,94 @@ class WelcomePageViewModelTest {
...
+    }
+
+    private fun emitShowDialogAndFireAndNeverHideTips() = coroutineRule.runBlocking {
```
[the remark this answers is not in the document]
Comment (anikiki) [reply app/src/androidTest/java/com/duckduckgo/app/onboarding/ui/page/WelcomePageViewModelTest.kt]: Not applicable anymore as I updated the tests.

File: statistics/src/main/java/com/duckduckgo/app/statistics/VariantManager.kt [hunk 10]
```diff
@@ -45,8 +48,13 @@ interface VariantManager {
             Variant(key = "se", weight = 0.0, features = emptyList(), filterBy = { isSerpRegionToggleCountry() }),
 
             // Fireproof experiment
-            Variant(key = "mi", weight = 1.0, features = emptyList(), filterBy = { isEnglishLocale() }),
-            Variant(key = "mj", weight = 1.0, features = listOf(VariantFeature.FireproofExperiment), filterBy = { isEnglishLocale() }),
+            Variant(key = "mi", weight = 0.0, features = emptyList(), filterBy = { isEnglishLocale() }),
+            Variant(key = "mj", weight = 0.0, features = listOf(VariantFeature.FireproofExperiment), filterBy = { isEnglishLocale() }),
```
Comment (anikiki) [inline statistics/src/main/java/com/duckduckgo/app/statistics/VariantManager.kt:52]: Weights for mi, mj were updated to 0.0 as per the discussion in the fireproof experiment.

File: app/src/main/res/layout/include_experiment_returning_users_cta.xml [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:id="@+id/cta"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:background="?toolbarBgColor"
```
[the remark this answers is not in the document]
Comment (anikiki) [reply app/src/main/res/layout/include_experiment_returning_users_cta.xml:24]: It is weird weird. This layout has to be as close as possible to the other ones, that's why you see that background. Will be addressed as part of the UI cleanup.

File: app/src/main/java/com/duckduckgo/app/cta/ui/Cta.kt (hunk 3 above)
Comment (anikiki) [reply app/src/main/java/com/duckduckgo/app/cta/ui/Cta.kt:374]: It's been added as there is a new layout that uses the same `showCta` function and may not have all these. Will be reverted after the experiment.

**Pull request merged**